### PR TITLE
Added Bubble Sort Algorithm

### PR DESCRIPTION
**Pull request by jthorat10**
#### Short Discription
Bubble sort is a simple sorting algorithm that repeatedly steps through the list, compares adjacent elements, and swaps them if they are in the wrong order.

#### Brief description of what is fixed or changed
Take an array of numbers "5 1 4 2 8", and sort the array from lowest number to greatest number using bubble sort. In each step, elements written in bold are being compared. Three passes will be required;

First Pass
( 5 1 4 2 8 ) → ( 1 5 4 2 8 ), Here, the algorithm compares the first two elements and swaps since 5 > 1.
( 1 5 4 2 8 ) → ( 1 4 5 2 8 ), Swap since 5 > 4
( 1 4 5 2 8 ) → ( 1 4 2 5 8 ), Swap since 5 > 2
( 1 4 2 5 8 ) → ( 1 4 2 5 8 ), Now, since these elements are already in order (8 > 5), the algorithm does not swap them.
Second Pass
( 1 4 2 5 8 ) → ( 1 4 2 5 8 )
( 1 4 2 5 8 ) → ( 1 2 4 5 8 ), Swap since 4 > 2
( 1 2 4 5 8 ) → ( 1 2 4 5 8 )
( 1 2 4 5 8 ) → ( 1 2 4 5 8 )

#### Other comments

REF LINK - [https://en.wikipedia.org/wiki/Bubble_sort](https://en.wikipedia.org/wiki/Bubble_sort)

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -8,3 +8,4 @@ Rajiv Ranjan Singh <rajivperfect007@gmail.com>
 Prashant Rawat <prashant.rawat216@gmail.com>
 Harsheet <harsheetkakar@gmail.com>
 Pratik Goyal <pratikgoyal2712@gmail.com>
+Jay Thorat <j.thorat10@gmail.com        >
```
Comment (czgdp1807) [inline AUTHORS:11]:
```suggestion
Jay Thorat <j.thorat10@gmail.com>
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
     cocktail_shaker_sort, quick_sort, longest_common_subsequence, is_ordered,
     upper_bound, lower_bound, longest_increasing_subsequence, next_permutation,
     prev_permutation)
+from pydatastructs.linear_data_structures.algorithms import bubble_sort
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:8]:
```suggestion
    prev_permutation, bubble_sort)
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 3]
```diff
@@ -1205,3 +1206,44 @@ def _prev_permutation_comp(x, y, _comp):
 
     return _permutation_util(array, start, end, comp,
                              _prev_permutation_comp)
+
+
+def bubble_sort(array,start = 0,):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1211]: See the API of other sorting algorithms and modify this accordingly.

Comment (jthorat10) [reply pydatastructs/linear_data_structures/algorithms.py:1211]: Yeah Sure!!

File: pydatastructs/linear_data_structures/algorithms.py [hunk 4]
```diff
@@ -1205,3 +1206,61 @@ def _prev_permutation_comp(x, y, _comp):
 
     return _permutation_util(array, start, end, comp,
                              _prev_permutation_comp)
+
+def bubble_sort(array, start = 0, **kwargs):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1210]:
```suggestion
def bubble_sort(array, **kwargs):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 5]
```diff
@@ -1205,3 +1206,61 @@ def _prev_permutation_comp(x, y, _comp):
...
+    >>> bubble_sort([3,5,4,1,2])
+    [1, 2, 3, 4, 5]
+    >>> bubble_sort([8,3,5,2,6,1,9],3)
+    [8, 3, 5, 1, 2, 6, 9]
+    >>> bubble_sort([6,9,3,2,5,4,0,7,1,8],3,end = 6)
+    [6, 9, 3, 0, 2, 4, 5, 7, 1, 8]
+    >>> bubble_sort([6,9,3,2,5,4,0,7,1,8],start = 3,end = 6)
+    [6, 9, 3, 0, 2, 4, 5, 7, 1, 8]
+    """
+    end = kwargs.get('end', len(array) - 1)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1257]:
```suggestion
    start = kwargs.get('start', 0)
    end = kwargs.get('end', len(array) - 1)
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 6]
```diff
@@ -1205,3 +1206,61 @@ def _prev_permutation_comp(x, y, _comp):
...
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    arr_len = len(array)
+    for i in range(arr_len - 1):
+        for j in range(start , end):
+            if _comp(array[j], array[j + 1],comp) is True :
+                continue
+            else:
+                array[j], array[j + 1] = array[j + 1], array[j]
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1265]:
```suggestion
            if _comp(array[j], array[j + 1],comp) is False:
                array[j], array[j + 1] = array[j + 1], array[j]
```

File: pydatastructs/linear_data_structures/algorithms.py (hunk 4 above)
Comment (jthorat10) [reply pydatastructs/linear_data_structures/algorithms.py]: Done!

File: pydatastructs/linear_data_structures/algorithms.py [hunk 7]
```diff
@@ -1205,3 +1206,58 @@ def _prev_permutation_comp(x, y, _comp):
...
+    >>> bubble_sort([6,9,3,2,5,4,0,7,1,8],start=3,end = 6)
+    [6, 9, 3, 0, 2, 4, 5, 7, 1, 8]
+    """
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    arr_len = len(array)
+    for i in range(arr_len - 1):
+        for j in range(start , end):
+            if _comp(array[j], array[j + 1],comp) is False :
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1261]:
```suggestion
            if _comp(array[j], array[j + 1], comp):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 8]
```diff
@@ -1205,3 +1206,58 @@ def _prev_permutation_comp(x, y, _comp):
...
+    >>> bubble_sort([6,9,3,2,5,4,0,7,1,8],start=3,end = 6)
+    [6, 9, 3, 0, 2, 4, 5, 7, 1, 8]
+    """
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    arr_len = len(array)
+    for i in range(arr_len - 1):
+        for j in range(start , end):
+            if _comp(array[j], array[j + 1], comp):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1261]:
```suggestion
            if not _comp(array[j], array[j + 1], comp):
```

**Comment by jthorat10** [comment]
This i causing due to that assertionError. I guess

**Comment by jthorat10** [comment]
`FAILED pydatastructs/linear_data_structures/tests/test_algorithms.py::test_bubble_sort - assert (15, 13, 31) == (12, 13, 31)`

Getting the same error in local tests too.

**Comment by czgdp1807** [comment]
Feel free to move ahead. Ignore this error. I will check on Sunday.

**Comment by jthorat10** [comment]
> Feel free to move ahead. Ignore this error. I will check on Sunday.

Sure!

File: pydatastructs/linear_data_structures/algorithms.py [hunk 9]
```diff
@@ -1205,3 +1206,60 @@ def _prev_permutation_comp(x, y, _comp):
...
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    arr_len = len(array)
+    for i in range(arr_len - 1):
+        for j in range(start , end):
+            if not _comp(array[j], array[j + 1], comp):
+                array[j], array[j + 1] = array[j + 1], array[j]
+    if _check_type(array, DynamicArray):
+        array._modify(force=True)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1264]: 👍🏻

File: pydatastructs/linear_data_structures/algorithms.py [hunk 10]
```diff
@@ -1205,3 +1206,60 @@ def _prev_permutation_comp(x, y, _comp):
 
     return _permutation_util(array, start, end, comp,
                              _prev_permutation_comp)
+
+def bubble_sort(array, **kwargs):
+    """
+    Sorts the provided array by bubble sort algorithm. In
+    which it checks consecutive two elements and if the prior
+    one is greater than the other then it swaps them and
+    returns sorted array
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1215]:
```suggestion
    Implements bubble sorts algorithm.
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 11]
```diff
@@ -1205,3 +1206,60 @@ def _prev_permutation_comp(x, y, _comp):
...
+
+    Examples
+    ========
+    >>> from pydatastructs import bubble_sort
+    >>> bubble_sort([3,5,4,1,2])
+    [1, 2, 3, 4, 5]
+    >>> bubble_sort([8,3,5,2,6,1,9],start = 3)
+    [8, 3, 5, 1, 2, 6, 9]
+    >>> bubble_sort([6,9,3,2,5,4,0,7,1,8],start=3,end = 6)
+    [6, 9, 3, 0, 2, 4, 5, 7, 1, 8]
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1253]: Can you just copy paste the examples from one of the sorting algorithms above?

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:1253]: Also add one references section (see merge_sort_parallel for an example).

File: pydatastructs/linear_data_structures/algorithms.py (hunk 9 above)
Comment (jthorat10) [reply pydatastructs/linear_data_structures/algorithms.py:1264]: Any changes or suggestions?

File: pydatastructs/linear_data_structures/algorithms.py [hunk 12]
```diff
@@ -1205,3 +1206,65 @@ def _prev_permutation_comp(x, y, _comp):
...
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    arr_len = len(array)
+    for i in range(arr_len - 1):
+        for j in range(start , end):
+            if not _comp(array[j], array[j + 1], comp):
+                array[j], array[j + 1] = array[j + 1], array[j]
+    if _check_type(array, DynamicArray):
+        array._modify(force=True)
+    return array
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1270]:
```suggestion
                array[j], array[j + 1] = array[j + 1], array[j]

    if _check_type(array, DynamicArray):
        array._modify(force=True)

    return array
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 13]
```diff
@@ -1205,3 +1206,65 @@ def _prev_permutation_comp(x, y, _comp):
...
+    '[1, 2, 3]'
+    >>> out = bubble_sort(arr, comp=lambda u, v: u > v)
+    >>> str(out)
+    '[3, 2, 1]'
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Bubble_sort
+    .. [2] https://www.javatpoint.com/bubble-sort
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1258]:
```suggestion
```

**Comment by czgdp1807** [comment]
Thanks a lot for working on this. I think we missed to add bubble sort to documentation website code. It would be great if you can make a follow up PR for adding a one line in https://github.com/codezonediitj/pydatastructs/blob/master/docs/source/pydatastructs/linear_data_structures/algorithms.rst. May be adding after, `.. autofunction:: pydatastructs.quick_sort` should work.

**Comment by jthorat10** [comment]
Yeah sure!!

**Pull request merged**